ref-arch/real-time-analytics.rst

```diff
@@ -0,0 +1,509 @@
+
+Expiring Old Data
+-----------------
+
+The rollups make queries faster but we still have a lot of raw data sitting around. How
+long you should keep each granularity of data is a business decision, but once you decide
+it's easy to write a function to expire old data:
+
+.. code-block:: plpgsql
+
+  -- another function for the master
+  CREATE FUNCTION expire_old_request_data RETURNS void
+  AS $$
+    SET citus.all_modification_commutative TO TRUE;
+    SELECT master_modify_multiple_shards(
+      'DELETE FROM http_request WHERE ingest_time < now() - interval ''1 hour'';');
+    SELECT master_modify_multiple_shards(
+      'DELETE FROM http_request_1min WHERE ingest_time < now() - interval ''1 day'';');
+    RESET citus.all_modification_commutative;
+  END;
+  $$ LANGUAGE 'sql';
+
+.. NOTE::
+
+  The above function should be called every minute. As mentioned above there are many
+  different ways to accomplish this and no way which makes everybody happy. If you're
+  capable of adding cron entries to the machine the master is running on you might
+  consider adding a crontab entry:
+
+  .. code-block:: bash
+  
+    * * * * * psql -c "SELECT expire_old_request_data();"
+
+Review, what have we done?
+--------------------------
+
+That's the entire architecture! The next few sections are just extensions for additional
+problems which often pop up. So what makes this better than what it replaced? Well, let's
+look again at the problem it solves. We wanted to enable a dashboard which aggregated:
+
+1. Large amounts of data with
+2. Low latency
+
+Where the naive solution struggled with a few problems:
+
+A. The dashboard must aggregate every row in the target date range for every query it
+   answers.
+B. Storage costs will grow proportionally with the ingest rate and the length of the
+   queryable history.
+
+Because we roll up the raw data, and the dashboard only runs queries on that raw data, it
+must do a constant amount of work for each user. Users who have much more visiters than
+average won't have a dashboard which works any slower. Because the rollups fire every
+minute,
+
+If you've heard of postgres' VACUUM, you know it can be a pain once you have a large
+number of rows. The write pattern we're using turns out to be the ideal VACUUM use-case.
+We never modify rows, we only write to one end of the table while deleting from the other
+end. When VACUUM runs it marks a `visibility map
+<https://www.postgresql.org/docs/9.5/static/storage-vm.html>`_ to keep track of which
+pages have already been vacuumed and can be skipped during the next vacuum. Since we never
+UPDATE, the only pages which have that bit reset are the pages of entirely DELETEd rows.
+All VACUUM needs to do is scan through and reclaim those empty pages, is happens very
+quickly!
+
+Approximate Distinct Counts
+---------------------------
+
+One kind of query we're particularily proud of is :ref:`approximate distinct counts
```

Comment:
"proud" seems a bit weird, maybe: "A common question in website analytics systems is: How many unique visitors ..." and later "we can answer the query approximately" -> "the HLL extension can answer the query approximately".